TalkAloud/TalkAloud/Managers/AudioManager.swift

```diff
@@ -133,28 +134,58 @@ class AudioManager {
     //==================================================
     
     public func setTag(for selectedRecording: AudioRecording, tag: String) {
+        
         CoreDataManager.sharedInstance.updateAudioRecordingTag(with: selectedRecording, with: tag)
     }
     
     public func removeTag(for selectedRecording: AudioRecording) {
         CoreDataManager.sharedInstance.removeAudioRecordingTag(for: selectedRecording)
     }
     
-    public func getAllAudioRecordingTags() -> [String]? {
-        var allTags = [String]()
-        
+    public func getTags(for selectedRecording: AudioRecording) -> [String] {
+        guard let tags = selectedRecording.tags else { return [] }
+        return tags
+    }
+    
+    public func getAllAudioRecordingTags() -> [TagModel]? {
         guard let audioRecordings = loadAllRecordings() else { return nil }
         
+        var allTags = [TagModel]()
+        
         for audioRecording in audioRecordings {
             if let tags = audioRecording.tags {
                 for tag in tags {
-                    allTags.append(tag)
+                    let tagModel = TagModel(tag: tag, isTagSelected: false)
+                    if !allTags.contains(tagModel) {
+                        allTags.append(tagModel)
+                    }
                 }
             }
         }
-        allUniqueTags = allTags.unique
+    
+        return allTags
+    }
+    
+    public func addATag(tagModel: TagModel, tagModels: [TagModel]) -> [TagModel] {
+        var tagModelDataSource = tagModels
+        
+        if !tagModelDataSource.contains(tagModel) {
+            tagModelDataSource.append(tagModel)
+        }
+        
+        return tagModelDataSource
+    }
+    
+    public func removeTags(tags: [String], tagModels: [TagModel]) -> [TagModel] {
```

Comment:
Also you shouldn't pass your dataSource from the ViewController to here. Modify this method so your AudioManager is the only source of truth. The data source from the AudioManager needs to "drive" the dataSource from the ViewController.